File: WinForms/.NET/FormDesigner/TextBoxPropertiesForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using O2S.Components.PDFView4NET.Forms;
using O2S.Components.PDFView4NET.Graphics.Fonts;
using O2S.Components.PDFView4NET.Graphics;
using O2S.Components.PDFView4NET.Annotations;
using O2S.Components.PDFView4NET;
using System.Drawing.Text;

namespace FormDesigner
{
    public partial class TextBoxPropertiesForm : Form
    {
        public TextBoxPropertiesForm()
        {
            InitializeComponent();
        }

        private PDFTextBoxWidget textBoxWidget;
        /// <summary>
        /// Gets or sets the textbox widget edited in this form.
        /// </summary>
        public PDFTextBoxWidget TextBoxWidget
        {
            get { return textBoxWidget; }
            set { textBoxWidget = value; }
        }

        private void TextBoxPropertiesForm_Load(object sender, EventArgs e)
        {
            LoadFonts();

            PDFTextBoxField textBoxField = textBoxWidget.Field as PDFTextBoxField;

            PDFFieldWidgetVisibilityStatus[] visibilityStatus = new PDFFieldWidgetVisibilityStatus[] 
                { 
                    PDFFieldWidgetVisibilityStatus.Visible, PDFFieldWidgetVisibilityStatus.Hidden, 
                    PDFFieldWidgetVisibilityStatus.VisibleNonPrintable, PDFFieldWidgetVisibilityStatus.HiddenPrintable 
                };
            cbxVisibility.SelectedIndex = Array.IndexOf<PDFFieldWidgetVisibilityStatus>(visibilityStatus, textBoxWidget.VisibilityStatus);
            cbxOrientation.SelectedIndex = ((int)textBoxWidget.Rotation) / 90;
            cbxAlign.SelectedIndex = (int)textBoxWidget.Align;

            int borderWidthSelectedIndex = -1;
            if (textBoxWidget.BorderWidth == 0)
            {
                borderWidthSelectedIndex = 0;
            }
            if (textBoxWidget.BorderWidth == 1)
            {
                borderWidthSelectedIndex = 1;
            }
            if (textBoxWidget.BorderWidth == 2)
            {
                borderWidthSelectedIndex = 2;
            }
            if (textBoxWidget.BorderWidth == 3)
            {
                borderWidthSelectedIndex = 3;
            }
            cbxBorderWidth.SelectedIndex = borderWidthSelectedIndex;
            PDFBorderStyle[] borderStyles = new PDFBorderStyle[]
                {
                    PDFBorderStyle.Solid, PDFBorderStyle.Dashed, PDFBorderStyle.Beveled, PDFBorderStyle.Inset, PDFBorderStyle.Underline
                };
            cbxBorderStyle.SelectedIndex = Array.IndexOf<PDFBorderStyle>(borderStyles, textBoxWidget.BorderStyle);

            if (textBoxWidget.BorderColor != null)
            {
                btnBorderColor.BackColor = textBoxWidget.BorderColor.ToGdiColor();
            }
            else
            {
                btnBorderColor.BackColor = Color.Transparent;
            }
            if (textBoxWidget.BackColor != null)
            {
                btnFillColor.BackColor = textBoxWidget.BackColor.ToGdiColor();
            }
            else
            {
                btnFillColor.BackColor = Color.Transparent;
            }
            if (textBoxWidget.TextColor != null)
            {
                btnTextColor.BackColor = textBoxWidget.TextColor.ToGdiColor();
            }
            else
            {
                btnTextColor.BackColor = Color.Transparent;
            }
            cbxFont.SelectedIndex = 0;
            cbxFontSize.SelectedIndex = 3;

            txtFieldName.Text = textBoxWidget.Field.Name;
            txtTooltip.Text = textBoxWidget.Field.Tooltip;

            chkMultiline.Checked = textBoxField.Multiline;
            chkScrollLongText.Checked = textBoxField.ScrollLongText;
            chkMaxLength.Checked = textBoxField.MaxLength > 0;
            chkPassword.Checked = textBoxField.Password;
            chkFileSelection.Checked = textBoxField.FileSelect;
            chkCheckSpelling.Checked = textBoxField.CheckSpelling;
            chkComb.Checked = textBoxField.Comb;

            chkReadOnly.Checked = textBoxField.ReadOnly;
            chkNoExport.Checked = textBoxField.NoExport;
            chkLocked.Checked = (textBoxWidget.Flags & PDFAnnotationFlags.Locked) == PDFAnnotationFlags.Locked;
        }

        private void cbxBorderWidth_SelectedIndexChanged(object sender, EventArgs e)
        {
            cbxBorderStyle.Enabled = cbxBorderWidth.SelectedIndex != 0;
            btnBorderColor.Enabled = cbxBorderWidth.SelectedIndex != 0;
        }

        private void btnSelectColor_Click(object sender, EventArgs e)
        {
            cd.Color = (sender as Button).BackColor;
            if (cd.ShowDialog() == DialogResult.OK)
            {
                (sender as Button).BackColor = cd.Color;
            }
        }

        private void chkMaxLength_CheckedChanged(object sender, EventArgs e)
        {
            txtMaxLength.Enabled = chkMaxLength.Checked;
            lblMaxLengthChars.Enabled = chkMaxLength.Checked;
            if (!chkMaxLength.Checked)
            {
                txtMaxLength.Text = "";
            }
        }

        private void chkComb_CheckedChanged(object sender, EventArgs e)
        {
            txtComb.Enabled = chkComb.Checked;
            lblCombChars.Enabled = chkComb.Checked;
            if (!chkComb.Checked)
            {
                txtComb.Text = "";
            }
        }

        private void btnOK_Click(object sender, EventArgs e)
        {
            PDFTextBoxField textBoxField = textBoxWidget.Field as PDFTextBoxField;
            if (textBoxField.Name != txtFieldName.Text)
            {
                // Multiple widgets, rename just the selected one.
                if (textBoxField.Widgets.Count > 1)
                {
                    PDFPage parentPage = textBoxWidget.Page;

                    // Current widget is renamed by creating a new field with the new name
                    // and assigning the selected widget to the new field.
                    // First remove the widget from current parent field.
                    textBoxField.Widgets.Remove(textBoxWidget);
                    // Create the "renamed" field, clear its default widget, 
                    // add the selected widget and then add the field to the page.
                    PDFTextBoxField newField = new PDFTextBoxField(txtFieldName.Text);
                    newField.Widgets.RemoveAt(0);
                    newField.Widgets.Add(textBoxWidget);
                    parentPage.Fields.Add(newField);

                    textBoxField = newField;
                }
                else
                {
                    textBoxField.Name = txtFieldName.Text;
                }
            }
            textBoxField.Tooltip = txtTooltip.Text;

            PDFFieldWidgetVisibilityStatus[] visibilityStatus = new PDFFieldWidgetVisibilityStatus[] 
                { 
                    PDFFieldWidgetVisibilityStatus.Visible, PDFFieldWidgetVisibilityStatus.Hidden, 
                    PDFFieldWidgetVisibilityStatus.VisibleNonPrintable, PDFFieldWidgetVisibilityStatus.HiddenPrintable 
                };
            textBoxWidget.VisibilityStatus = visibilityStatus[cbxVisibility.SelectedIndex];
            textBoxWidget.Rotation = (O2S.Components.PDFView4NET.PDFRotationMode)(cbxOrientation.SelectedIndex * 90);
            textBoxWidget.Align = (PDFFieldTextAlign)cbxAlign.SelectedIndex;
            textBoxWidget.BorderWidth = cbxBorderWidth.SelectedIndex;
            textBoxWidget.BorderColor = new PDFRgbColor(btnBorderColor.BackColor);
            PDFBorderStyle[] borderStyles = 
                (PDFBorderStyle[])Enum.GetValues(typeof(PDFBorderStyle));
            textBoxWidget.BorderStyle = (PDFBorderStyle)cbxBorderStyle.SelectedIndex;
            textBoxWidget.BackColor = new PDFRgbColor(btnFillColor.BackColor);

            textBoxWidget.TextColor = new PDFRgbColor(btnTextColor.BackColor);
            textBoxWidget.Font = GetSelectedFont();

            textBoxField.Multiline = chkMultiline.Checked;
            textBoxField.ScrollLongText = chkScrollLongText.Checked;
            int charCount = 0;
            int.TryParse(txtMaxLength.Text, out charCount);
            if (charCount > 0)
            {
                textBoxField.MaxLength  = charCount;
            }
            textBoxField.Password = chkPassword.Checked;
            textBoxField.FileSelect = chkFileSelection.Checked;
            textBoxField.CheckSpelling = chkCheckSpelling.Checked;
            int.TryParse(txtComb.Text, out charCount);
            if (charCount > 0)
            {
                textBoxField.Comb = chkComb.Checked;
                textBoxField.MaxLength = charCount;
            }

            textBoxField.ReadOnly = chkReadOnly.Checked;
            textBoxField.NoExport = chkNoExport.Checked;
            if (chkLocked.Checked)
            {
                textBoxWidget.Flags = textBoxWidget.Flags | PDFAnnotationFlags.Locked;
            }
            else
            {
                textBoxWidget.Flags = textBoxWidget.Flags & ~PDFAnnotationFlags.Locked;
            }

        }

        private void LoadFonts()
        {
            //cbxFont.Items.Add("------------------------------");
            //// Enumerate the fonts installed on local machine.
            //InstalledFontCollection ifc = new InstalledFontCollection();
            //FontFamily[] ff = ifc.Families;
            //for (int i = 0; i < ff.Length; i++)
            //{
            //    cbxFont.Items.Add(ff[i].Name);
            //    if (ff[i].IsStyleAvailable(FontStyle.Bold))
            //    {
            //        cbxFont.Items.Add(ff[i].Name + "- Bold");
            //    }
            //    if (ff[i].IsStyleAvailable(FontStyle.Italic))
            //    {
            //        cbxFont.Items.Add(ff[i].Name + "- Italic");
            //    }
            //    if (ff[i].IsStyleAvailable(FontStyle.Bold | FontStyle.Italic))
            //    {
            //        cbxFont.Items.Add(ff[i].Name + "- BoldItalic");
            //    }
            //}
        }

        private PDFFontBase GetSelectedFont()
        {
            PDFFontBase font = null;
            int fontIndex = cbxFont.SelectedIndex / 4;

            if (fontIndex < 3)
            {
                PDFFontFace[] faces = new PDFFontFace[] { PDFFontFace.Helvetica, PDFFontFace.TimesRoman, PDFFontFace.Courier };
                font = new PDFFont(faces[fontIndex], double.Parse(cbxFontSize.SelectedItem.ToString()));
                int fontStyle = cbxFont.SelectedIndex % 4;
                font.Bold = (fontStyle == 1) || (fontStyle == 3);
                font.Italic = (fontStyle == 2) || (fontStyle == 3);
            }
            return font;
        }
    }
}

File: WinForms/.NET Framework/FormDesigner/VBNET/TextBoxPropertiesForm.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports O2S.Components.PDFView4NET.Forms
Imports O2S.Components.PDFView4NET.Graphics.Fonts
Imports O2S.Components.PDFView4NET.Graphics
Imports O2S.Components.PDFView4NET.Annotations
Imports O2S.Components.PDFView4NET
Imports System.Drawing.Text

Namespace O2S.Samples.PDFView4NET.FormDesigner
	Public Partial Class TextBoxPropertiesForm
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private m_textBoxWidget As PDFTextBoxWidget
		''' <summary>
		''' Gets or sets the textbox widget edited in this form.
		''' </summary>
		Public Property TextBoxWidget() As PDFTextBoxWidget
			Get
				Return m_textBoxWidget
			End Get
			Set
				m_textBoxWidget = value
			End Set
		End Property

        Private Sub TextBoxPropertiesForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load

            Dim textBoxField As PDFTextBoxField = TryCast(m_textBoxWidget.Field, PDFTextBoxField)

            Dim visibilityStatus As PDFFieldWidgetVisibilityStatus() = New PDFFieldWidgetVisibilityStatus() {PDFFieldWidgetVisibilityStatus.Visible, PDFFieldWidgetVisibilityStatus.Hidden, PDFFieldWidgetVisibilityStatus.VisibleNonPrintable, PDFFieldWidgetVisibilityStatus.HiddenPrintable}
            cbxVisibility.SelectedIndex = Array.IndexOf(Of PDFFieldWidgetVisibilityStatus)(visibilityStatus, m_textBoxWidget.VisibilityStatus)
            cbxOrientation.SelectedIndex = CInt(m_textBoxWidget.Rotation) / 90
            cbxAlign.SelectedIndex = CInt(m_textBoxWidget.Align)

            Dim borderWidthSelectedIndex As Integer = -1
            If m_textBoxWidget.BorderWidth = 0 Then
                borderWidthSelectedIndex = 0
            End If
            If m_textBoxWidget.BorderWidth = 1 Then
                borderWidthSelectedIndex = 1
            End If
            If m_textBoxWidget.BorderWidth = 2 Then
                borderWidthSelectedIndex = 2
            End If
            If m_textBoxWidget.BorderWidth = 3 Then
                borderWidthSelectedIndex = 3
            End If
            cbxBorderWidth.SelectedIndex = borderWidthSelectedIndex
            Dim borderStyles As PDFBorderStyle() = New PDFBorderStyle() {PDFBorderStyle.Solid, PDFBorderStyle.Dashed, PDFBorderStyle.Beveled, PDFBorderStyle.Inset, PDFBorderStyle.Underline}
            cbxBorderStyle.SelectedIndex = Array.IndexOf(Of PDFBorderStyle)(borderStyles, m_textBoxWidget.BorderStyle)

            If m_textBoxWidget.BorderColor IsNot Nothing Then
                btnBorderColor.BackColor = m_textBoxWidget.BorderColor.ToGdiColor()
            Else
                btnBorderColor.BackColor = Color.Transparent
            End If
            If m_textBoxWidget.BackColor IsNot Nothing Then
                btnFillColor.BackColor = m_textBoxWidget.BackColor.ToGdiColor()
            Else
                btnFillColor.BackColor = Color.Transparent
            End If
            If m_textBoxWidget.TextColor IsNot Nothing Then
                btnTextColor.BackColor = m_textBoxWidget.TextColor.ToGdiColor()
            Else
                btnTextColor.BackColor = Color.Transparent
            End If
            cbxFont.SelectedIndex = 0
            cbxFontSize.SelectedIndex = 3

            txtFieldName.Text = m_textBoxWidget.Field.Name
            txtTooltip.Text = m_textBoxWidget.Field.Tooltip

            chkMultiline.Checked = textBoxField.Multiline
            chkScrollLongText.Checked = textBoxField.ScrollLongText
            chkMaxLength.Checked = textBoxField.MaxLength > 0
            chkPassword.Checked = textBoxField.Password
            chkFileSelection.Checked = textBoxField.FileSelect
            chkCheckSpelling.Checked = textBoxField.CheckSpelling
            chkComb.Checked = textBoxField.Comb

            chkReadOnly.Checked = textBoxField.[ReadOnly]
            chkNoExport.Checked = textBoxField.NoExport
            chkLocked.Checked = (m_textBoxWidget.Flags And PDFAnnotationFlags.Locked) = PDFAnnotationFlags.Locked
        End Sub

        Private Sub cbxBorderWidth_SelectedIndexChanged(ByVal sender As Object, ByVal e As EventArgs) Handles cbxBorderWidth.SelectedIndexChanged
            cbxBorderStyle.Enabled = cbxBorderWidth.SelectedIndex <> 0
            btnBorderColor.Enabled = cbxBorderWidth.SelectedIndex <> 0
        End Sub

        Private Sub btnSelectColor_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnBorderColor.Click, btnFillColor.Click, btnTextColor.Click
            cd.Color = TryCast(sender, Button).BackColor
            If cd.ShowDialog() = DialogResult.OK Then
                TryCast(sender, Button).BackColor = cd.Color
            End If
        End Sub

        Private Sub chkMaxLength_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chkMaxLength.CheckedChanged
            txtMaxLength.Enabled = chkMaxLength.Checked
            lblMaxLengthChars.Enabled = chkMaxLength.Checked
            If Not chkMaxLength.Checked Then
                txtMaxLength.Text = ""
            End If
        End Sub

        Private Sub chkComb_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles chkComb.CheckedChanged
            txtComb.Enabled = chkComb.Checked
            lblCombChars.Enabled = chkComb.Checked
            If Not chkComb.Checked Then
                txtComb.Text = ""
            End If
        End Sub

        Private Sub btnOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnOK.Click
            Dim textBoxField As PDFTextBoxField = TryCast(m_textBoxWidget.Field, PDFTextBoxField)
            If textBoxField.Name <> txtFieldName.Text Then
                ' Multiple widgets, rename just the selected one.
                If textBoxField.Widgets.Count > 1 Then
                    Dim parentPage As PDFPage = m_textBoxWidget.Page

                    ' Current widget is renamed by creating a new field with the new name
                    ' and assigning the selected widget to the new field.
                    ' First remove the widget from current parent field.
                    textBoxField.Widgets.Remove(m_textBoxWidget)
                    ' Create the "renamed" field, clear its default widget, 
                    ' add the selected widget and then add the field to the page.
                    Dim newField As New PDFTextBoxField(txtFieldName.Text)
                    newField.Widgets.RemoveAt(0)
                    newField.Widgets.Add(m_textBoxWidget)
                    parentPage.Fields.Add(newField)

                    textBoxField = newField
                Else
                    textBoxField.Name = txtFieldName.Text
                End If
            End If
            textBoxField.Tooltip = txtTooltip.Text

            Dim visibilityStatus As PDFFieldWidgetVisibilityStatus() = New PDFFieldWidgetVisibilityStatus() {PDFFieldWidgetVisibilityStatus.Visible, PDFFieldWidgetVisibilityStatus.Hidden, PDFFieldWidgetVisibilityStatus.VisibleNonPrintable, PDFFieldWidgetVisibilityStatus.HiddenPrintable}
            m_textBoxWidget.VisibilityStatus = visibilityStatus(cbxVisibility.SelectedIndex)
            m_textBoxWidget.Rotation = cbxOrientation.SelectedIndex * 90
            m_textBoxWidget.Align = DirectCast(cbxAlign.SelectedIndex, PDFFieldTextAlign)
            m_textBoxWidget.BorderWidth = cbxBorderWidth.SelectedIndex
            m_textBoxWidget.BorderColor = New PDFRgbColor(btnBorderColor.BackColor)
            Dim borderStyles As PDFBorderStyle() = DirectCast([Enum].GetValues(GetType(PDFBorderStyle)), PDFBorderStyle())
            m_textBoxWidget.BorderStyle = DirectCast(cbxBorderStyle.SelectedIndex, PDFBorderStyle)
            m_textBoxWidget.BackColor = New PDFRgbColor(btnFillColor.BackColor)

            m_textBoxWidget.TextColor = New PDFRgbColor(btnTextColor.BackColor)
            m_textBoxWidget.Font = GetSelectedFont()

            textBoxField.Multiline = chkMultiline.Checked
            textBoxField.ScrollLongText = chkScrollLongText.Checked
            Dim charCount As Integer = 0
            Integer.TryParse(txtMaxLength.Text, charCount)
            If charCount > 0 Then
                textBoxField.MaxLength = charCount
            End If
            textBoxField.Password = chkPassword.Checked
            textBoxField.FileSelect = chkFileSelection.Checked
            textBoxField.CheckSpelling = chkCheckSpelling.Checked
            Integer.TryParse(txtComb.Text, charCount)
            If charCount > 0 Then
                textBoxField.Comb = chkComb.Checked
                textBoxField.MaxLength = charCount
            End If

            textBoxField.[ReadOnly] = chkReadOnly.Checked
            textBoxField.NoExport = chkNoExport.Checked
            If chkLocked.Checked Then
                m_textBoxWidget.Flags = m_textBoxWidget.Flags Or PDFAnnotationFlags.Locked
            Else
                m_textBoxWidget.Flags = m_textBoxWidget.Flags And Not PDFAnnotationFlags.Locked
            End If

        End Sub

		Private Function GetSelectedFont() As PDFFontBase
			Dim font As PDFFontBase = Nothing
			Dim fontIndex As Integer = cbxFont.SelectedIndex / 4

			If fontIndex < 3 Then
				Dim faces As PDFFontFace() = New PDFFontFace() {PDFFontFace.Helvetica, PDFFontFace.TimesRoman, PDFFontFace.Courier}
				font = New PDFFont(faces(fontIndex), Double.Parse(cbxFontSize.SelectedItem.ToString()))
				Dim fontStyle As Integer = cbxFont.SelectedIndex Mod 4
				font.Bold = (fontStyle = 1) OrElse (fontStyle = 3)
				font.Italic = (fontStyle = 2) OrElse (fontStyle = 3)
			End If
			Return font
		End Function
	End Class
End Namespace
